### PR TITLE
preview

### PR DESCRIPTION
**Pull request by schweitzpgi**

### CODE REVIEW

File: flang/lib/Lower/ConvertExpr.cpp [hunk 1]
```diff
@@ -1697,133 +1625,486 @@ class ExprLowering {
...
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Expression traversal and lowering.
+  //===--------------------------------------------------------------------===//
+
+  // Lower the expression in a scalar context.
+  template <typename A>
+  ExtValue asScalar(const A &x) {
+    return ExprLowering{getLoc(), converter, symMap, stmtCtx}.genval(x);
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:1776]: I am not following by which magic this is computing the scalar outside of the loop (obviously, it is computed outside of the loops given your tests, so it works :) )

Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:1776]: Ah, found it ! it is not being generated in a lamda :)

**Review by jeanPerier — APPROVED** [review]
Kudos, I really enjoyed reading at ConvertExpr.cpp! The continuation style passing really keeps the complexity down. I think it also looks very solid in terms of what kind of array expression it will be able to handle.
The only thing that I think might hit some problems is the cleanup. The overall statement approach is tempting but I fear it might bumps in subtle issues when the statement lowering generate control flow that may (especially when it some times does, but sometimes does not).
But there are probably ways to ensure that the clean-up at the statement level will work. I guess some experience with real code is needed here.

File: flang/lib/Lower/StatementContext.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+#include "SymbolMap.h"
+#include "llvm/Support/ErrorHandling.h"
+#include <functional>
+
+namespace Fortran::lower {
+
+/// When lowering a statement, large temporaries may be allocated on the heap to
+/// buffer intermediate results. These temporaries must be deallocated at the
+/// end of the statement. These deallocations are threaded via a
+/// StatementContext back to the "end" of the statement.
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/StatementContext.h:25]: Yep. We'll have to be very careful in order to get the cleanups in the correct positions. (Consider what I've done just a very rough draft in that regard. I expect it's buggy.)

Also I called it a "statement" context for convenience. It's really much more like a synchronization point in C. So we'll have to lower something like
```fortran
  IF (exp) 100, 200, 300
```
as
```
   eval exp
   cleanup any temps
   switch <, 100 ; =, 200 ; >,300
```
and not
```
   eval exp
   switch <, 100 ; =, 200 ; >,300
   cleanup any temps
```
obviously. :)

File: flang/lib/Lower/ConvertExpr.cpp [hunk 3]
```diff
@@ -1697,133 +1625,486 @@ class ExprLowering {
...
-    std::visit([&](const auto &e) { genpre(e); }, exp.u);
+  static bool isArray(const A &x) {
+    return x.Rank() != 0;
   }
 
   template <typename A, typename = std::enable_if_t<Fortran::common::HasMember<
                             A, Fortran::evaluate::TypelessExpression>>>
-  void genpre(const A &) {}
+  CC genarr(const A &x) {
+    auto result = asScalar(x);
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:1788]: Hmm. I think this might be left over from some older work here.

**Pull request merged**